Library/OcCpuLib/OcCpuLib.c

```diff
@@ -1105,8 +1105,8 @@ ScanIntelProcessor (
 
   //
   // Calculate number of cores
-  //
-  if (Cpu->MaxId >= CPUID_CACHE_PARAMS && Cpu->Model <= CPU_MODEL_PENRYN) {
+  // If we are under visualization, then we should get topology from acpi as Penryn
+  if ((Cpu->MaxId >= CPUID_CACHE_PARAMS && Cpu->Model <= CPU_MODEL_PENRYN) || Cpu->Hypervisor) {
```

Comment:
At the very least, the condition must be corrected to "MaxId && (Penryn || HV)". As you mentioned MSRs may be problematic, the else-branch would need to be adapted too. However for Intel, I have not heard of issues yet.
@vit9696 Care to comment?

Reply:
You comments about the branches is right, I will change it later.
And the Intel problem is real, check the issues I mentioned, when setup QEMU smp and using newer model than Penryn it will failed.

Reply:
To me it looks good enough if we fix the braces and ignore the else part.